Replace `option.Config.{Get,Set,Append}Devices` by table lookups

bimmlerd opened the pull request:
This replaces usages of the Get/SetDevices APIs by lookups to the device table, and includes the necessary plumbing of that everywhere.

Scary stuff includes the wireguard interface thing in daemon/loader, not sure my transformation is functionally equivalent. CI will hopefully tell.

bimmlerd committed:
dp/iptables: remove usages of GetDevices

GetDevices returns a snapshot, but devices can change at runtime. The
new, table-based API reflects this fact. This patch, however, does not
change the semantics - it merely replaces the usage of GetDevices with a
single table lookup, to allow us to react to changes in the future (and
remove the GetDevices API).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
linux/node: remove usages of GetDevices

GetDevices returns a snapshot, but devices can change at runtime. The
new, table-based API reflects this fact. This patch, however, does not
change the semantics - it merely replaces the usage of GetDevices with a
single table lookup, to allow us to react to changes in the future (and
remove the GetDevices API).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
linux/config: remove usages of GetDevices

GetDevices returns a snapshot, but devices can change at runtime. The
new, table-based API reflects this fact. This patch, however, does not
change the semantics - it merely replaces the usage of GetDevices with a
single table lookup, to allow us to react to changes in the future (and
remove the GetDevices API).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
daemon: move devices flag to device controller

Since the devices controller is really the owner of this flag, move the
configuration into the cell proper, to avoid other things being able to
depend on it.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
daemon: remove usages of GetDevices

GetDevices returns a snapshot, but devices can change at runtime. The
new, table-based API reflects this fact. This patch, however, does not
change the semantics - it merely replaces the usage of GetDevices with a
single table lookup, to allow us to react to changes in the future (and
remove the GetDevices API).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
config: deprecate GetDevices/SetDevices

Devices (i.e. network interfaces) can change at runtime. The current
device API, however, suggest that these devices never change. Indeed, a
large portion of the agent is written with that assumption (though
workarounds have been added where changes to the devices broke things).

As part of an ongoing process to migrate the agent towards a future
where a changing set of devices is the norm, deprecate the existing,
snapshot-oriented APIs and replace it with table lookups.

In subsequent commits, usage of this API is reduced, and replaced with
usage of an API which allows for being notified when the set of devices
changes.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
dp/linux: use hive for node test setup

This is handy for future commits in which we increase the number of
dependencies we'd have to manually construct when creating a linux
NodeHandler.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
dp/linux: cleanups for linux node tests

* Refactor out the duplicated code to assert against NeighList
* Use proper fake nodemap in TestNodeUpdateIDs

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
linux/config: remove unneeded bwmanager reference

There are no usages, hence it is safe to remove.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
dp/loader: remove usages of GetDevices

GetDevices returns a snapshot, but devices can change at runtime. The
new, table-based API reflects this fact. This patch, however, does not
change the semantics - it merely replaces the usage of GetDevices with a
single table lookup, to allow us to react to changes in the future (and
remove the GetDevices API).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
endpoint: use fake datapath in tests

While the tests claim they need the full datapath, that is no longer
true. Instead, provide a fake datapath (to avoid having to provide all
the dependencies).

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

joamaki commented on pkg/datapath/loader/cache_test.go:
```diff
@@ -102,7 +117,19 @@ func (s *LoaderTestSuite) TestobjectCacheParallel(c *C) {
 
 		results := make(chan buildResult, t.builds)
-		cache := newObjectCache(&config.HeaderfileWriter{}, nil, tmpDir)
+		devices := statedb.MustNewTable[*tables.Device]("devices", tables.DeviceIDIndex, tables.DeviceNameIndex, tables.DeviceSelectedIndex)
```
Use `NewDeviceTable`

joamaki commented on pkg/datapath/loader/cache_test.go:
```diff
@@ -23,7 +25,20 @@ func (s *LoaderTestSuite) TestobjectCache(c *C) {
 	defer cancel()
 
-	cache := newObjectCache(&config.HeaderfileWriter{}, nil, tmpDir)
+	devices := statedb.MustNewTable[*tables.Device]("devices", tables.DeviceIDIndex, tables.DeviceNameIndex, tables.DeviceSelectedIndex)
```
Use `NewDeviceTable`

joamaki commented on pkg/datapath/loader/template_test.go:
```diff
@@ -21,10 +23,21 @@ func (s *LoaderTestSuite) TestWrap(c *C) {
 	realEP := testutils.NewTestEndpoint()
 	template := wrap(&realEP, nil)
-	cfg := &config.HeaderfileWriter{}
+	devices := statedb.MustNewTable[*tables.Device]("devices", tables.DeviceIDIndex, tables.DeviceNameIndex, tables.DeviceSelectedIndex)
```
Use `NewDeviceTable`

bimmlerd committed:
dmn: clarify choice of loading bpf_{network,host}

The agent needs to choose between loading bpf_network or bpf_host onto
physical devices. In the long-term, we want to get rid of bpf_network.
In the short-term, however, it's still around, so we need to keep the
existing semantics intact.

Previous code chose by checking the devices configuration (i.e.
--devices) as well as the runtime-derived list of
devices, to understand whether we were auto-detecting devices. If so,
bpf_network would not be loaded in reinitializeIPSec, as bpf_host would
be loaded.

With the change towards always auto detecting devices, these checks need
to be switched for AreDevicesRequired in the short-term, a dedicated
flag in the long term to allow clean fall-back to encryptInterface and
the removal of bpf_network in the longer term.

Co-authored-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
config: Remove GetDevices and SetDevices

The snapshot based APIs are impossible to use correctly, as they do not
reflect the fact that the set of devices can change at runtime. Uses of
this API have been replaced with lookups into the Table[Device] in the
past few commits, hence removal is now easy.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
linux: include WG for EP static data and host DP

As explained in 0670fd2ace1136a1143a384b39912f8153d7a063, the WireGuard
device must be considered for loading bpf_host in certain conditions.

With the removal of the Get/Set/AppendDevices APIs, the global addition
of the WG interface to devices is no longer possible, nor was it
ever really desirable.

Since we don't want to make WG a "selected" device, we move the
inclusion of this interface to the two places necessary: when writing
static EP data in the config writer, as well as the loading of bpf_host.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

lmb commented on pkg/datapath/linux/config/config.go:
```diff
@@ -1035,13 +1034,27 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
+	nativeDevices, _ := tables.SelectedDevices(h.devices, h.db.ReadTxn())
+	deviceNames := tables.DeviceNames(nativeDevices)
+
+	// When WG & encrypt-node are on, a NodePort BPF to-be forwarded request
```
Why is the headerfile writer the right place for this judicious hack? Could it live elsewhere?

bimmlerd replied:
We need some of the static data to be written here (otherwise we'll fail with some errors like `symbol 'ETH_HLEN' not found`)

I think the truly clean solution would be to have a separate table to know which programs want to be loaded on which devices, but we're not there yet. I think this is better that the original hack, which was just adding it to the devices globally.

bimmlerd replied:
@joamaki you wanted to check with @brb regarding this - lmk of the resolution

lmb commented on pkg/datapath/linux/config/config.go:
```diff
@@ -99,6 +105,9 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	extraMacrosMap := make(dpdef.Map)
 	cDefinesMap := make(dpdef.Map)
 
+	txn := h.db.ReadTxn()
```
StateDB noob: when does this txn close?

joamaki replied:
ReadTxn's don't need to be closed.

lmb commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -297,38 +304,49 @@ func createVlanLink(vlanId int, mainLink *netlink.Dummy, c *C) *netlink.Vlan {
 	}()
 
+	// Define set of vlans which we want to allow.
+	allow := map[int]struct{}{
```
Nit: `map[int]bool` might be easier to work with?

bimmlerd replied:
Hm, not that it matters here, but map[]struct{} doesn't allocate for the values - I always thought it's the canonical way of declaring a Set in Go

lmb replied:
It is! I just find `map[foo]bool` much more ergonomic in cases where the memory doesn't matter.

bimmlerd replied:
done :)

lmb commented on pkg/datapath/loader/base.go:
```diff
@@ -307,9 +310,13 @@ func (l *loader) reinitializeXDPLocked(ctx context.Context, extraCArgs []string)
 // and reinsertion of the object into the kernel as well as an atomic program replacement
 // at the XDP hook. extraCArgs can be passed-in in order to alter BPF code defines.
 func (l *loader) ReinitializeXDP(ctx context.Context, o datapath.BaseProgramOwner, extraCArgs []string) error {
+	// TODO: react to changes (using the currently ignored watch channel)
```
Not a big fan of TODOs. Something that needs to happen in this PR? Something that needs a follow up issue?

bimmlerd replied:
Fair enough. It won't happen in this PR and it's tracked in an issue in a different repo - do you prefer not having the TODO in that case?

lmb replied:
Yes, but that is just a nit.

lmb commented on pkg/datapath/loader/cache_test.go:
```diff
@@ -23,7 +25,23 @@ func (s *LoaderTestSuite) TestobjectCache(c *C) {
 	defer cancel()
 
-	cache := newObjectCache(&config.HeaderfileWriter{}, nil, tmpDir)
+	devices, err := tables.NewDeviceTable()
```
Nit: pull this into a helper?

bimmlerd replied:
Ideally, this would be using hive instead, but this PR is big enough as is - sure, I'll put it into a helper.

lmb commented on pkg/datapath/loader/loader.go:
```diff
@@ -526,8 +532,22 @@ func (l *loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
+		nativeDevices, _ := tables.SelectedDevices(l.devices, l.db.ReadTxn())
+		devices := tables.DeviceNames(nativeDevices)
+
+		// When WG & encrypt-node are on, a NodePort BPF to-be forwarded request
```
Hmm this seems familiar!

bimmlerd replied:
:) yeah, it's not exactly DRY. Do you have a suggestion on how to not do this in both places? I don't want to add to `config.Option` (indeed, working to the contrary), but it's not super clear to me which component should know about this. I was suprised in the first place that adding it here wasn't enough (due to the compilation failure due to `ETH_HLEN`)

lmb replied:
I'm not qualified enough to suggest a better place unfortunately.

derailed commented on daemon/cmd/config.go:
```diff
@@ -153,7 +154,8 @@ func getConfigHandler(d *Daemon, params GetConfigParams) middleware.Responder {
 
 	// Manually add fields that are behind accessors.
-	m["Devices"] = option.Config.GetDevices()
+	devs, _ := tables.SelectedDevices(d.devices, d.db.ReadTxn())
```
Fetching device names seems to be repeated thru out this pr. ie get selected devices -> extract device names. Perhaps best to make it a convenience call aka `tables.SelectedDeviceNames`?

bimmlerd replied:
I want to avoid encouraging dumbing down the device structs to strings as much as possible - it's as is to avoid blowing up this PR more. Subsequent PRs will transform more of the downstream code to take a device struct, so then this will go away.

derailed commented on pkg/datapath/linux/config/config.go:
```diff
@@ -1035,13 +1034,27 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
 }
 
-func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.EndpointConfiguration) error {
+func (h *HeaderfileWriter) writeTemplateConfig(devices []*tables.Device, fw *bufio.Writer, e datapath.EndpointConfiguration) error {
```
nit: Not a hard rule here but writeXXX with a writer would be best if writer figures as first arg.

derailed reviewed (commented):
@bimmlerd Nice work David!

tommyp1ckles commented on daemon/cmd/config.go:
```diff
@@ -153,7 +154,8 @@ func getConfigHandler(d *Daemon, params GetConfigParams) middleware.Responder {
 
 	// Manually add fields that are behind accessors.
-	m["Devices"] = option.Config.GetDevices()
+	devs, _ := tables.SelectedDevices(d.devices, d.db.ReadTxn())
```
Since we're not checking the invalidate channel, the device table is being passed in with the daemonParams, is there any risk of another module changing the device table giving us a possibly ambiguous `m["Devices"]`value?

bimmlerd replied:
Yeah, that's the issue all of this current code has. This PR doesn't solve that problem (which existed already), but at least allows doing it - subsequent PRs will do the reconcilation part.

tommyp1ckles commented on pkg/datapath/loader/template_test.go:
```diff
@@ -21,10 +23,24 @@ func (s *LoaderTestSuite) TestWrap(c *C) {
 	realEP := testutils.NewTestEndpoint()
 	template := wrap(&realEP, nil)
-	cfg := &config.HeaderfileWriter{}
+	devices, err := tables.NewDeviceTable()
```
nit: there's going to be a lot of this boilerplate code moving forward if we need to setup the statedb fixtures manually each time.  Would we want to start converting tests to just use Hive which could just provide the statedb.Cell etc...?

bimmlerd replied:
I wanted to avoid making this PR even bigger - I pulled it into a helper for now.

tommyp1ckles reviewed (approved):
Endpoint changes look good.
